jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1027,54 +1057,68 @@ public void run()
             try
             {
                 Runnable job = null;
-                while (true)
+                exit: while (true)
                 {
                     // If we had a job,
                     if (job != null)
                     {
+                        // signal that we are idle again,
                         idle = true;
-                        // signal that we are idle again
                         if (!addCounts(0, 1))
                             break;
                     }
-                    // else check we are still running
+                    // else check we are still running.
                     else if (_counts.getHi() == Integer.MIN_VALUE)
                     {
                         break;
                     }
 
                     try
                     {
-                        // Look for an immediately available job
-                        job = _jobs.poll();
+                        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+                        job = idleJobPoll(idleTimeoutNanos);
                         if (job == null)
                         {
-                            // No job immediately available maybe we should shrink?
-                            long idleTimeout = getIdleTimeout();
-                            if (idleTimeout > 0 && getThreads() > _minThreads)
+                            // No job available after an idle timeout, maybe we should shrink?
+                            if (idleTimeoutNanos > 0 && getThreads() > getMinThreads())
                             {
-                                long last = _lastShrink.get();
                                 long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
+                                long shrinkPeriod = idleTimeoutNanos / getIdleTimeoutMaxShrinkCount();
+                                if (LOG.isDebugEnabled())
+                                    LOG.debug("Shrink check, period={}ms {}", TimeUnit.NANOSECONDS.toMillis(shrinkPeriod), QueuedThreadPool.this);
+                                while (true)
                                 {
-                                    if (LOG.isDebugEnabled())
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
-                                    break;
+                                    long lastShrink = _lastShrink.get();
+                                    long prevShrink = lastShrink;
+                                    // If the shrink window is too far in the past,
+                                    // advance it to be one idle timeout before now.
```

Comment:
I don't like using the word window, as we don't fully embrace that abstraction. how about:
```suggestion
                                    // If the shrink threshold is too far in the past,
                                    // set it to just one idle timeout before now.
```

Reply:
Uhm, it is a window between some time in the past and the current time.
And it is a credit-based scheme like the HTTP/2 flow control window, every expired thread reduces the window until zero, and the passing of time gives it more credit.

Reply:
I understand what a window system is and how it is an analogy to the algorithm here.
I just do not think that in this case it adds any clarity by describing it as a window.

Best just to say when the next shrink is allowed to happen.

Reply:
> I just do not think that in this case it adds any clarity by describing it as a window.

But it's not a threshold either, and calling it so it's even more confusing.

> Best just to say when the next shrink is allowed to happen.

The value is completely unrelated to "when the next shrink is allowed to happen", so I cannot say that.

I don't understand why you're so against the term "window", it is perfectly representing what it is: a value that has a fixed maximum, can decrement until it closes down to zero, and when it closes, no more air (err shrink), but can reopen 😃 

We can call it `shrinkCredit` if you want, but seems to me unnecessary wording when "window" is precisely describing it.
"Allowance"? "Quota"?

If we change the math to use a counter instead of a nanoTime, how would you call the counter? 
There would be N "shrinkCredits" and each time a thread is eligible to shrink you would decrement by 1 credit.
As time passes you gain 1 credit for every shrink period that has passed.

Reply:
A window has two ends that adjust and can't be allowed to get too far apart.  This just doesn't have that abstraction.  It is kind of similar, but not close enough that the term window introduces confusion rather than clarity.
It only works for you because you have implemented a window with a single counter (window size), but the origin of the term really depends on having two pointers and a window between them that is limited in size.  

It is not a quota, nor allowance, nor credit.   It is a time that is related to now, that moves forward in time and acts as a threshold to determine if an expiry is allowed or skipped.